docs: add plugin limits and lifecycle sections (#12697)
hashicorp/vault: website/content/docs/internals/plugins.mdx

built-in plugin. In such a situation, Vault will always choose the custom plugin
when enabling it.

## Plugin Lifecycle

Vault plugins are long-running processes that remain running once they are
spawned by Vault, the parent process. Plugin processes can be started by Vault's
active node and performance standby nodes. Additionally, there are cases where

> plugin processes may be terminated by Vault. These cases include the following:

Reviewer: ```suggestion
plugin processes may be terminated by Vault. These cases include but
are not limited to:
```

Author: Updated in d1979de.

- Vault active node step-down
- Vault barrier seal
- Vault graceful shutdown
- Database configured connection update
- Database root credentials rotation

> The lifecycle of plugin processes should be managed by automatically and

Reviewer: Some more suggested wordsmithing:
```
The lifecycle of plugin processes is managed automatically by Vault. Termination
of these processes is typical in certain scenarios (such as those listed above).
Vault will start plugin processes when needed, typically by lazily loading the
plugin when a request that requires the plugin is received by Vault. A plugin
process may be started or terminated through other internal processes
within Vault as well.
```

Reviewer: Should we also include a note about how a plugin processes should not be terminated by anything other than Vault?

Author: Updated in d1979de, and also added a note on out-of-band termination.

Reviewer: What do you think about changing the first sentence to say "...plugin processes are managed by Vault." I'm namely thinking of using less passive language - "is managed" vs "should be managed".
